libass/ass_alloc.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2020 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "config.h"
+#include "ass_compat.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "ass_alloc.h"
+#include "ass_utils.h"
+
+// strdup and strndup are only part of ISO C since C2x; libass only requires C99
+#ifndef HAVE_STRDUP
+char *ass_strdup_replacement(const char *str)
+{
+    size_t len    = strlen(s) + 1;
+    char *new_str = ass_malloc(len);
+    if (new_str)
+        memcpy(new_str, str, len);
+    return new_str;
+}
+#define strdup ass_strdup_replacement
+#endif
+
+#ifndef HAVE_STRNDUP
+char *ass_strndup_replacement(const char *s, size_t n)
+{
+    char *end = memchr(s, 0, n);
+    size_t len = end ? end - s : n;
+    char *new = len < SIZE_MAX ? ass_malloc(len + 1) : NULL;
+    if (new) {
+        memcpy(new, s, len);
+        new[len] = 0;
+    }
+    return new;
+}
+#define strndup ass_strndup_replacement
+#endif
+
+
+// Following macros expect file, line and lib to be defined and initialised
+#ifndef NDEBUG
+    #define ASS_ALLOC_TEST(ptr) \
+        if (!ptr) \
+            ass_msg(lib, MSGL_ERR, "Alloc fail in %s at line %u!\n", file, line)
+#else
+    #define ASS_ALLOC_TEST(ptr) do { } while (0)
```

Comment:
Sorry, I don't quite follow. Are you suggesting, there should be an `ass_nomem_log` implementation (without force) which will only be used in `NDEBUG` builds?

Reply:
Yes, the whole reporting thing started due to lack of messages in case of malloc failure. Testing/fuzzing/debugging is only secondary. There should be `ass_nomem_log()` without force (or `ASS_NO_MEM()` macro with automatic message). And it's implementation can do full detailed reporting in debug mode and use flag and common message for release.